keychain_win.cpp

```diff
@@ -0,0 +1,211 @@
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+// clang-format off
+// make sure windows.h is included before wincred.h
+#include "keychain.h"
+
+#include <memory>
+
+#define UNICODE
+
+#include <windows.h>
+#include <wincred.h>
+#include <intsafe.h> // for DWORD_MAX
+// clang-format on
+
+namespace {
+
+static DWORD cred_type = CRED_TYPE_GENERIC;
+
+struct LpwstrDeleter {
+    void operator()(WCHAR *p) const { delete[] p; }
+};
+using ScopedLpwstr = std::unique_ptr<WCHAR, LpwstrDeleter>;
+
+LPWSTR utf8ToWideChar(const std::string &utf8) {
+    int len = MultiByteToWideChar(
+        CP_UTF8,
+        0, // flags must be 0 for UTF-8
+        utf8.c_str(),
+        -1,      // rely on null-terminated input string
+        nullptr, // no out-buffer needed
+        0); // return required buffer size rather than writing to an out-buffer
+    if (len == 0) {
+        return nullptr;
+    }
+
+    LPWSTR buffer = new WCHAR[len];
+    int bytesWritten =
+        MultiByteToWideChar(CP_UTF8, 0, utf8.c_str(), -1, buffer, len);
+
+    if (bytesWritten == 0) {
+        // failure; cleanup buffer manually
+        delete[] buffer;
+        return nullptr;
+    }
+
+    return buffer;
+}
+
+std::string wideCharToAnsi(LPWSTR wChar) {
+    std::string result;
+    if (wChar == nullptr) {
+        return result;
+    }
+
+    int len =
+        WideCharToMultiByte(CP_ACP, 0, wChar, -1, nullptr, 0, nullptr, nullptr);
+    if (len == 0) {
+        return result;
+    }
+
+    std::unique_ptr<char[]> buffer(new char[len]);
+    int bytesWritten = WideCharToMultiByte(
+        CP_ACP, 0, wChar, -1, buffer.get(), len, nullptr, nullptr);
+
+    if (bytesWritten != 0) {
+        result = std::string(buffer.get());
+    }
+
+    return result;
+}
+
+std::string getErrorMessage(DWORD errorCode) {
+    std::string errMsg;
+    LPWSTR errBuffer = nullptr;
+    ::FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM,
+                    nullptr, // ignored for the flags we use
+                    errorCode,
+                    0, // figure out LANGID automatically
+                    reinterpret_cast<LPWSTR>(&errBuffer),
+                    0,        // figure out out-buffer size automatically
+                    nullptr); // no additional arguments
+    if (errBuffer != nullptr) {
+        errMsg = wideCharToAnsi(errBuffer);
+        LocalFree(errBuffer);
+    }
+    return errMsg;
+}
+
+std::string makeTargetName(const std::string &package,
+                           const std::string &service,
+                           const std::string &user) {
+    return package + "." + service + '/' + user;
+}
```

Comment:
The result of `makeTargetName` is always directly transformed via `utf8ToWideChar` and then stored in a `ScopedLpwstr`. You could hide those implementation details in this helper as well.